zerver/lib/test_runner.py

```diff
@@ -451,18 +429,51 @@ def run_tests(self, test_labels, extra_tests=None,
         try:
             suite = self.build_suite(test_labels, extra_tests)
         except AttributeError:
-            # We are likely to get here only when running a single test
-            # suite on Python 3.4 or lower.
+            # We are likely to get here only when running tests in serial
+            # mode on Python 3.4 or lower.
             traceback.print_exc()
             print()
             print("  This is often caused by a test module/class/function that doesn't exist or ")
-            print("  import properly. You can usually debug in a `manage.py shell` via e.g. ")
+            print("  import properly. You can usually debug in a `./manage.py shell` via e.g. ")
             print("    import zerver.tests.test_messages")
             print("    from zerver.tests.test_messages import StreamMessagesTest")
             print("    StreamMessagesTest.test_message_to_stream")
             print()
             sys.exit(1)
 
+        # Checking for import failures in serial mode
+        if self.parallel == 1 and hasattr(suite, '_tests') and len(suite._tests):
+            if hasattr(suite._tests[0], '_exception'):
+                print()
+                print(suite._tests[0]._exception)
+                sys.exit(1)
+
+        # Checking for import failures in parallel mode
+        elif hasattr(suite, 'subsuites'):
+            # When a test suite has failed imports, test_name usually has one
+            # of the following forms:
+            # unittest.loader.ModuleImportFailure.zerver.tests.test_upload
+            # (on Python 3.4 or lower) or
+            # unittest.loader._FailedTest.zerver.tests.test_upload
+            # (on Python 3.5 or higher).
+            import_failure_prefix_old = 'unittest.loader.ModuleImportFailure.'
+            import_failure_prefix_new = 'unittest.loader._FailedTest.'
+            for subsuite in suite.subsuites:
+                if not isinstance(subsuite, tuple) or len(subsuite) < 2:
+                    continue
+                test_names = subsuite[1]
+                if not isinstance(test_names, list):
+                    continue
+                for test_name in test_names:
```

Comment:
This loop should work for both parallel and serial models.

Reply:
Could you show me how the code would look like? I haven't found a way to do that.

Reply:
See the diff above.